File: sample-projects/php/src/Domain/Service/CreatureService.php
<?php

declare(strict_types=1);

namespace De\Sots\CellarsAndCentaurs\Domain\Service;

use De\Sots\CellarsAndCentaurs\Domain\Model\Creature;
use De\Sots\CellarsAndCentaurs\Domain\Model\CreatureId;
use Psr\Log\LoggerInterface;

final class CreatureService
{
    public function __construct(private Creatures $creatures, private LoggerInterface $logger)
    {
    }

    public function save(Creature $creature): void
    {
        $this->logger->info('Saving creature ' . $creature->getId()->idAsString());
        $this->creatures->save($creature);
    }

    public function find(CreatureId $id): Creature
    {
        return $this->creatures->find($id);
    }
}

File: sample-projects/c/src/domain/service/creature_service.c
#include "../../../include/de/sots/cellarsandcentaurs/domain/service/creature_service.h"
#include "../../../include/de/sots/cellarsandcentaurs/domain/service/creatures.h"
#include "../../../include/de/sots/cellarsandcentaurs/domain/model/creature.h"

CreatureService creature_service_new(Creatures *creatures)
{
    CreatureService service = { creatures };
    return service;
}

void creature_service_save(CreatureService *service, Creature *creature)
{
    service->creatures->save(service->creatures->self, creature);
}
